rmw_fastrtps_cpp/src/publisher.cpp

```diff
@@ -0,0 +1,180 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include "fastrtps/Domain.h"
+#include "fastrtps/participant/Participant.h"
+
+#include "rmw/allocators.h"
+#include "rmw/error_handling.h"
+#include "rmw/rmw.h"
+
+#include "rmw_fastrtps_shared_cpp/create_rmw_gid.hpp"
+#include "rmw_fastrtps_shared_cpp/custom_participant_info.hpp"
+#include "rmw_fastrtps_shared_cpp/custom_publisher_info.hpp"
+#include "rmw_fastrtps_shared_cpp/names.hpp"
+#include "rmw_fastrtps_shared_cpp/namespace_prefix.hpp"
+#include "rmw_fastrtps_shared_cpp/qos.hpp"
+#include "rmw_fastrtps_shared_cpp/rmw_common.hpp"
+
+#include "rmw_fastrtps_cpp/identifier.hpp"
+#include "rmw_fastrtps_cpp/publisher.hpp"
+
+#include "type_support_common.hpp"
+
+using Domain = eprosima::fastrtps::Domain;
+using Participant = eprosima::fastrtps::Participant;
+using TopicDataType = eprosima::fastrtps::TopicDataType;
+
+rmw_publisher_t *
+rmw_fastrtps_cpp::create_publisher(
+  const CustomParticipantInfo * participant_info,
+  const rosidl_message_type_support_t * type_supports,
+  const char * topic_name,
+  const rmw_qos_profile_t * qos_policies,
+  const rmw_publisher_options_t * publisher_options,
+  bool keyed,
+  bool create_publisher_listener)
+{
+  if (!participant_info) {
+    RMW_SET_ERROR_MSG("participant_info is null");
+    return nullptr;
+  }
+  if (!topic_name || strlen(topic_name) == 0) {
+    RMW_SET_ERROR_MSG("publisher topic is null or empty string");
+    return nullptr;
+  }
+  if (!qos_policies) {
+    RMW_SET_ERROR_MSG("qos_policies is null");
+    return nullptr;
+  }
+  if (!publisher_options) {
+    RMW_SET_ERROR_MSG("publisher_options is null");
+    return nullptr;
+  }
+
+  const rosidl_message_type_support_t * type_support = get_message_typesupport_handle(
+    type_supports, RMW_FASTRTPS_CPP_TYPESUPPORT_C);
+  if (!type_support) {
+    type_support = get_message_typesupport_handle(
+      type_supports, RMW_FASTRTPS_CPP_TYPESUPPORT_CPP);
+    if (!type_support) {
+      RMW_SET_ERROR_MSG("type support not from this implementation");
+      return nullptr;
+    }
+  }
+
+  CustomPublisherInfo * info = nullptr;
+  rmw_publisher_t * rmw_publisher = nullptr;
+  eprosima::fastrtps::PublisherAttributes publisherParam;
+
+  if (!is_valid_qos(*qos_policies)) {
+    return nullptr;
+  }
+
+  // Load default XML profile.
+  Domain::getDefaultPublisherAttributes(publisherParam);
+
+  // TODO(karsten1987) Verify consequences for std::unique_ptr?
```

Comment:
@Karsten1987 what is this about? I assume this was copied from somewhere, but still I'm curious.

Reply:
I moved that from `rmw_publisher.cpp`.
I needed to be able to create a publisher without a `node`, that's why the moved code.

Reply:
Still curious what this TODO is about. @Karsten1987

Reply:
I should have written a bit more context to this for my future self. I don't have a real good answer to do, but I assume it was trying to use `unique_ptr` instead of `new` in the code below. I guess that could eliminate the `goto fail` part.

Reply:
Since it isn't using `unique_ptr`, yet could we just remove the TODO? @ivanpauno I'll let you decide and act on it or resolve this.

Reply:
Deleted those TODOs.
I will open an issue explaining some possible "cpp style" enhacements that includes that, as I described in another comment https://github.com/ros2/rmw_fastrtps/pull/312#discussion_r375378060.